src/pip/_internal/index.py

```diff
@@ -939,15 +939,22 @@ def _get_encoding_from_headers(headers):
     return None
 
 
-_CLEAN_LINK_RE = re.compile(r'[^a-z0-9$&+,/:;=?@.#%_\\|-]', re.I)
+_CLEAN_LINK_RE = re.compile(r"^(/[a-z])%3A", re.I)
 
 
 def _clean_link(url):
     # type: (str) -> str
     """Makes sure a link is fully encoded.  That is, if a ' ' shows up in
     the link, it will be rewritten to %20 (while not over-quoting
     % or other characters)."""
-    return _CLEAN_LINK_RE.sub(lambda match: '%%%2x' % ord(match.group(0)), url)
+    result = urllib_parse.urlparse(url)
+    # Quote URL while making sure we are not double quoting anything.
+    result = result._replace(path=urllib_parse.quote(
+        urllib_parse.unquote(result.path)))
+    # Unquote anything that looks like a quoted drive letter.
+    result = result._replace(path=_CLEAN_LINK_RE.sub(r"\1:", result.path))
```

Comment:
Why replace twice? I would just do something like:

```python
path = urllib_parse.quote(urllib_parse.unquote(result.path)))
result = result._replace(path=_CLEAN_LINK_RE.sub(r"\1:", .path))
```

(Although there is likely no significant impact either way.)

Reply:
That does look more readable :smile:. Done